CDAP-7634 First cut of AuthEnforce annotation and class rewrite

rohitsinha54 opened the pull request:
- First cut of AuthEnforce annotation and class rewrite to generate Authorization enforcement call.
- Currently, the rewrite supports only EntityId from method parameters and not multiple part string or looking into class fields. 
- One thing is which is different from the design is that in design we mentioned entity being specified in AuthEnforce annotation will be name of the variable from the method parameter or class field. Supporting this is not possible as after compilation method parameter names are lost and they are represented/accessed by position. So we expect use to mark parameters with Name annotation and give unique name to them. Will update the design to reflect this.
- Have added extensive unit-tests to cover all possible scenarios in which the AuthEnforce annotation can be used. Please provide comments if I have missed a case.

Design: https://wiki.cask.co/display/CE/Authorization+Improvements

Build: http://builds.cask.co/browse/CDAP-DUT5031-9
IT:

rohitsinha54 commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java:
```diff
@@ -303,7 +304,8 @@ public synchronized void delete(final NamespaceId namespaceId) throws Exception
+  @AuthEnforce(entity = "namespaceId", enforceOn = NamespaceId.class, privileges = Action.ADMIN)
+  public synchronized void deleteDatasets(@Name("namespaceId") NamespaceId namespaceId) throws Exception {
     // TODO: CDAP-870, CDAP-1427: Delete should be in a single transaction.
     if (!exists(namespaceId)) {
```
@chtyim : The authorization enforcement line is generated as the first line in this method. I think we should be doing the existence checks before auth checks as before. In this cause should we create a deleteDatasetsHelper method which does the delete after auth check and keep the pre checks here ? Or is there a better way to do this ?

rohitsinha54 commented on cdap-common/src/main/java/co/cask/cdap/common/app/MainClassLoader.java:
```diff
@@ -111,22 +116,37 @@ public static MainClassLoader createFromContext(FilterClassLoader.Filter filter)
+      // in co.cask.cdap package
+      //TODO (Rohit) : This is a hack. Figure out why ClassNotFound is thrown for Hbase AlreadyExistsException
+      rewritesNeeded.setAuthRewriteNeeded(className.startsWith("co.cask.cdap") &&
+                                            !className.equalsIgnoreCase("co.cask.cdap.data2.util.hbase" +
```
Note: HbaseTableUtil filter will be removed before merge.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,243 @@
+/**
+ * A {@link ClassRewriter} for rewriting bytecode of classes which needs Authorization Enforcement and uses
+ * {@link AuthEnforce} annotation.
+ * This class rewriter first visit every method of the class to check if it had {@link AuthEnforce} annotation.
```
Please write the javadoc in a way the shows how generated code looks like. See `DatasetClassRewriter` and `HttpHandlerGenerator` (for more complicated case) as an example.

Nishith commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java:
```diff
@@ -318,7 +320,6 @@ public synchronized void deleteDatasets(NamespaceId namespaceId) throws Exceptio
     }
 
     // Namespace data can be deleted. Revoke all privileges first
```
No need for this.

Nishith commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,243 @@
+
+      // We use AdviceAdapter to process the AuthEnforce annotation and insert authorization enforcement check as the
+      // fist line in the method
+      return new AdviceAdapter(Opcodes.ASM5, mv, access, name, desc) {
```
Please expand on the comments explaining how this visits, the annotation and extracts all the parameters from it.

rohitsinha54 commented:
Closing this PR for now. Will open with classloading fix.

Commit:
CDAP-7634 First cut of AuthEnforce annotation and class rewrite

rohitsinha54 commented:
@chtyim : Fixed the classloading issue with two pass as discussed in-person. Also did the refactoring to call a static method which does the authorization. Please have a look. Thanks.

rohitsinha54 commented:
Build passed here: http://builds.cask.co/browse/CDAP-DUT5031-11

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/app/MainClassLoader.java:
```diff
@@ -152,4 +169,28 @@ protected boolean needIntercept(String className) {
     return urls.toArray(new URL[urls.size()]);
   }
+
+  private final class RewritesNeeded {
```
static class.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/app/MainClassLoader.java:
```diff
@@ -152,4 +169,28 @@ protected boolean needIntercept(String className) {
   }
+
+  private final class RewritesNeeded {
+    boolean datasetRewriteNeeded;
```
private field.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/app/MainClassLoader.java:
```diff
@@ -152,4 +169,28 @@ protected boolean needIntercept(String className) {
+    void setAuthRewriteNeeded(boolean authRewriteNeeded) {
+      this.authRewriteNeeded = authRewriteNeeded;
+    }
+    boolean needsRewrite() {
```
Missing line separation.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforce.java:
```diff
@@ -0,0 +1,56 @@
+public @interface AuthEnforce {
+
+  // For details of these fields see class documentation
+  String[] entities();
```
Please add javadoc for all the methods on this annotation.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforce.java:
```diff
@@ -0,0 +1,56 @@
+  // For details of these fields see class documentation
+  String[] entities();
+
+  Class enforceOn();
```
This should be `Class<? extends EntityId>`?

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceAnnotationEnforcer.java:
```diff
@@ -0,0 +1,49 @@
+// from the package because after class rewrite AuthEnforce annotations are rewritten to make call to the methods
+// in this class and since AuthEnforce annotation can be in any package after class rewrite this class methods
+// might be being called from other packages.
+public class AuthEnforceAnnotationEnforcer {
```
This should be a util class. Make it final and have private constructor. Also, call it `AuthEnforceUtil`.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceAnnotationEnforcer.java:
```diff
@@ -0,0 +1,49 @@
+   * @throws Exception
+   */
+  public static void enforce(AuthorizationEnforcer authorizationEnforcer, EntityId entities,
+                             AuthenticationContext authenticationContext, Set<Action> actions) throws Exception {
```
This method brings an interesting case. Since this method has `throws Exception`, meaning the caller method needs to have `throws Exception` as well. It used to be enforced by the Java compiler, however, since now we are generating code at runtime, it's possible for one to write a method like this:

``` java
@AuthEnforce(...)
public void someFunc(...) {
   // do something
}
```

We better test this case, as I am not sure how the JVM will handle it.

rohitsinha54 replied:
Nice observation. I added one unit test for this case and it works in current implementation.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,449 @@
+ * {@link AuthEnforce} annotation.
+ * </p>
+ * <p>
+ * This class rewriter does two pass on the class:
```
passes

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,449 @@
+    // SKIP_CODE to make the first pass faster since in the first pass we just want to process annotations to check
+    // if the class has any method with AuthEnforce annotation. If such method is found we also store the parameters
+    // which has the named annotations as specified in the entities field of the AuthEnforce.
+    cr.accept(new AuthEnforceAnnotationVisitor(className, cw), ClassReader.SKIP_CODE);
```
Should skip everything that can be skipped: `SKIP_CODE | SKIP_DEBUG | SKIP_FRAMES`

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,449 @@
+    // dependent classes for a class and this might lead to class loading problems if some dependent classes are not
+    // available. For example HbaseTableUtil use Hbase classes and in in-memory and standalone cdap hbase classes
+    // are not present.
+    ClassWriter cw = new ClassWriter(0);
```
Why need a class writer in the first pass?

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,449 @@
+  @Override
+  public byte[] rewriteClass(String className, InputStream input) throws IOException {
+    byte[] classBytes = ByteStreams.toByteArray(input);
+    methodAnnotations = new HashMap<>(); // re-initialize for this new class.
```
Better not use field. Have a method in `AuthEnforceAnnotationVisitor` class to return this map instead.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,449 @@
+        return mv;
+      }
+
+      final boolean[] hasAuthEnforce = new boolean[1];
```
Why not just make this (and the next) fields in the `MethodVisitor` you are creating below?

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,449 @@
+        @Override
+        public AnnotationVisitor visitAnnotation(String desc, boolean visible) {
+          AnnotationVisitor annotationVisitor = super.visitAnnotation(desc, visible);
+          if (visible) {
```
Better do

``` java
if (!visible) {
  return av;
}
...
```

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,449 @@
+          AnnotationVisitor annotationVisitor = super.visitAnnotation(desc, visible);
+          if (visible) {
+            // if the annotation is present then visit the annotation
+            if (AuthEnforce.class.getName().equals(Type.getType(desc).getClassName())) {
```
Similar to above, invert the condition and return early.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,449 @@
+                @Override
+                public void visit(String name, Object value) {
+                  try {
+                    enforceOn[0] = Class.forName(((Type) value).getClassName());
```
You don't need to load the `Class`, right?? You just need the asm `Type`, isn't it?

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/app/MainClassLoader.java:
```diff
@@ -111,22 +116,34 @@ public static MainClassLoader createFromContext(FilterClassLoader.Filter filter)
+      rewritesNeeded.setDatasetRewriteNeeded(Classes.isSubTypeOf(className, DATASET_CLASS_NAME, resourceLookup, cache));
+      // Authorization annotation can only exists in cdap classes so we need to overwrite only classes
+      // in co.cask.cdap package
+      rewritesNeeded.setAuthRewriteNeeded(className.startsWith("co.cask.cdap"));
```
Should be `co.cask.cdap.` (the dot at the end).

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/app/MainClassLoader.java:
```diff
@@ -49,8 +52,10 @@
+  private final AuthEnforceRewriter authEnforceRewriter;
   private final Function<String, URL> resourceLookup;
   private final Map<String, Boolean> cache;
+  private RewritesNeeded rewritesNeeded;
```
Using this field is a bit unsafe, since classloading can be concurrent if concurrent classloading attribute is enabled.

chtyim replied:
Maybe better just call `Classes.isSubTypeOf` again in the `rewriteClass` method (the decision is cached in the `cache` map, so the overhead is super small, also it's easy to just turn the map to concurrent map if we decided to enable parallel classloading).

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforce.java:
```diff
@@ -0,0 +1,65 @@
+@Target(ElementType.METHOD)
+public @interface AuthEnforce {
+
+  // Specifies the entity on which authorization will be enforced. Method parameters and class variables which are
```
Use javadoc style, not inline comment style.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,446 @@
+    // and parameters for the method for second pass in which class rewrite will be performed.
+    ClassReader cr = new ClassReader(classBytes);
+
+    // SKIP_CODE to make the first pass faster since in the first pass we just want to process annotations to check
```
update comment.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,446 @@
+              ClassReader.SKIP_CODE | ClassReader.SKIP_DEBUG | ClassReader.SKIP_FRAMES);
+
+    Map<Method, AnnotationDetails> methodAnnotations = classVisitor.getMethodAnnotations();
+    if (!methodAnnotations.isEmpty()) {
```
Flip the condition, it'll be easier to read.

``` java
if (methodAnnotations.isEmpty()) {
  // No need to rewrite
  return classBytes;
}
// Do the rewrite stuff.
```

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,446 @@
+   * method with {@link AuthEnforce} annotations. Note: This is a visitor, to know the order of in which the below
+   * overridden methods will be called and their overall responsibility please see {@link ClassVisitor} documentation.
+   */
+  private final class AuthEnforceAnnotationVisitor extends ClassVisitor {
```
static class.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,446 @@
+
+    private final String className;
+    private boolean interfaceClass;
+    private final Map<Method, AnnotationDetails> methodAnnotations;
```
Have all `final` field comes before non-final fields.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,446 @@
+      // all the AuthEnforce annotation details through AnnotationVisitor and also sets a boolean flag hasEnforce
+      // which is used later in visitParameterAnnotation to visit the annotations on parameters of this method only
+      // if it had AuthEnforce annotation.
+      mv = new MethodVisitor(Opcodes.ASM5, mv) {
```
You can just `return new MethodVisitor(....) { ... }`.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,446 @@
+        return mv;
+      }
+
+
```
Do you like extra new line?

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,446 @@
+        public AnnotationVisitor visitAnnotation(String desc, boolean visible) {
+          AnnotationVisitor annotationVisitor = super.visitAnnotation(desc, visible);
+
+
```
Do you like extra new line?

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,446 @@
+          // We found AuthEnforce annotation so visit the annotation to collect all the specified details
+          // See AnnotationVisitor documentation to know the order in which the below overridden methods are
+          // called and their responsibilities.
+          annotationVisitor = new AnnotationVisitor(Opcodes.ASM5, annotationVisitor) {
```
I think you can simplify this by using `AnnotationNode` instead. Do we support multiple `@AuthEnforce` annotations on the same method? If not, then it can simply be:

``` java
new MethodVisitor(...) {
  AnnotationNode authEnforceNode;   // if we support multiple @AuthEnforce, then this becomes a List<>

  @Override
  public AnnotationVisitor visitAnnotation(String desc, boolean visible) {
    if (!(visible && AuthEnforce.class.getName().equals(Type.getType(desc).getClassName()))) {
     return super.visitAnnotation(desc, visible);
    }
    authEnforceNode = new AnnotationNode(desc);
    return authEnforceNode;
  }
}
```

chtyim replied:
Then you can use the `authEnforceNode.values` in the `visitEnd` method to extracts the attributes on the `@AuthEnforce` annotation.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,446 @@
+
+        @Override
+        public AnnotationVisitor visitParameterAnnotation(final int parameter, String desc, boolean visible) {
+          AnnotationVisitor annotationVisitor = super.visitParameterAnnotation(parameter, desc, visible);
```
Similar to above, I believe the code would be simpler if you use `AnnotationNode` as well.

chtyim replied:
No need to call super.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,446 @@
+   * A wrapper to store all the annotation details of method which has {@link AuthEnforce} annotation and parameters
+   * marked with {@link Name} annotation specifying the entities
+   */
+  private class AnnotationDetails {
```
You can combine this and the `AuthAnnotation` class, right? Also, make it final static class.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,446 @@
+                                     String[] exceptions) {
+      MethodVisitor mv = super.visitMethod(access, methodName, methodDesc, signature, exceptions);
+      // From the first pass we know the methods which has AuthEnforce annotation and needs to be rewritten
+      if (methodAnnotations.containsKey(new Method(methodName, methodDesc))) {
```
Flip the condition and return early.

chtyim replied:
Also, you can just call get.

``` java
AnnotationDetails details = methodAnnotations.get(...);
if (details == null) {
  return mv;
}
return new AdviceAdapter(...) {
...
}
```

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,446 @@
+            // Get the annotation details which was collected by visiting annotations during the first pass
+            AnnotationDetails annotationDetails = methodAnnotations.get(new Method(methodName, methodDesc));
+
+            LOG.debug("AuthEnforce annotation found in class {} on method {}. Authorization enforcement command will " +
```
Make this `trace`, otherwise you'll see tons of this message when log level is DEBUG.

rohitsinha54 commented:
@chtyim : Addressed all the comments in the "Addressed comments 2" commit. Also, added field and its setter and updated unit tests for the same in most recent last two commits. Please have a look. Thanks.

rohitsinha54 commented:
@chtyim : Also added support for specifying class fields in AuthEnforce in last commit in this PR itself rather than a new one since this PR is still open.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,555 @@
+ * Before:
+ * <pre>
+ *     public class ValidAuthEnforceAnnotations {
+ *      private final AuthorizationEnforcer authorizationEnforcer;
```
Before shouldn't have these two fields.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,555 @@
+ * After:
+ * <pre>
+ *     public class ValidAuthEnforceAnnotations {
+ *      private final AuthorizationEnforcer authorizationEnforcer;
```
The field name should be generated

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,555 @@
+ *      private final AuthenticationContext authenticationContext;
+ *
+ *      public ValidAuthEnforceAnnotations() {
+ *        authenticationContext = new AuthenticationTestContext();
```
The setting should be done through `@Inject` setter method, not in the constructor.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,555 @@
+ *
+ *      public void testSingleAction(@Name("namespaceId") NamespaceId namespaceId) throws Exception {
+ *        AuthEnforceUtil.enforce(this.authorizationEnforcer, namespaceId, authenticationContext,
+ *                                              Set<Action.Admin>);
```
The sample code is not valid.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,555 @@
+    private final String className;
+    private final Map<Method, AnnotationDetails> methodAnnotations;
+    private boolean interfaceClass;
+    private final Map<String, Type> fieldDetails;
```
Group final fields together.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,555 @@
+   * A wrapper to store all the annotation details of method which has {@link AuthEnforce} annotation and parameters
+   * marked with {@link Name} annotation specifying the entities
+   */
+  private static final class AnnotationDetails {
```
Detail, no s

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,555 @@
+            getField(classType, fieldNameWithoutThis, fieldDetails.get(fieldNameWithoutThis));
+          } else if (annotationDetails.getParameterAnnotation().containsKey(name)) { // method parameter has preference
+            // + 1 because 0 specify "this"
+            visitVarInsn(ALOAD, annotationDetails.getParameterAnnotation().get(name) + 1);
```
Are you loading method argument? If it is, better use `loadArg`.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,555 @@
+    @Override
+    public FieldVisitor visitField(int access, final String fieldName, String desc, String signature, Object value) {
+      fieldDetails.put(fieldName, Type.getType(desc));
+      return super.visitField(access, fieldName, desc, signature, value);
```
You can `return null`, if the class is an interface. This skips field visiting

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,555 @@
+      // No need to visit the class to look for AuthEnforce annotation if this class is an interface or the method is
+      // static.
+      if (interfaceClass || Modifier.isStatic(access)) {
+        return mv;
```
Similar to above, you can `return null` to skip method visitor to speed up. This also means no need to call `super.visitMethod`.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,555 @@
+
+          // If the annotation is not visible or its not AuthEnforce just skip
+          if (!(visible && AuthEnforce.class.getName().equals(Type.getType(desc).getClassName()))) {
+            return annotationVisitor;
```
Same here, `null` can be returned.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,555 @@
+    @Override
+    public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
+      // visit the header of the class to identify if this class is an interface
+      super.visit(version, access, name, signature, superName, interfaces);
```
No need to call super.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,555 @@
+
+        @Override
+        public AnnotationVisitor visitAnnotation(String desc, boolean visible) {
+          AnnotationVisitor annotationVisitor = super.visitAnnotation(desc, visible);
```
No need to call super.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,555 @@
+        public AnnotationVisitor visitParameterAnnotation(final int parameter, String desc, boolean visible) {
+          AnnotationVisitor annotationVisitor = super.visitParameterAnnotation(parameter, desc, visible);
+          if (!(hasAuthEnforce && visible && Name.class.getName().equals(Type.getType(desc).getClassName()))) {
+            return annotationVisitor;
```
Can return `null`.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,555 @@
+      return new MethodVisitor(Opcodes.ASM5, mv) {
+
+        final Map<Integer, AnnotationNode> parameterAnnotationNode = new HashMap<>();
+        boolean hasAuthEnforce;
```
No need to have this boolean. The `authEnforceAnnotationNode` will be `null` if the method is not annotated.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,555 @@
+  private static class AuthEnforceAnnotationNodeProcessor {
+
+    final Set<Action> actions = new HashSet<>();
+    Type enforceOn = null;
```
No need to assign to `null`.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,555 @@
+
+    final Set<Action> actions = new HashSet<>();
+    Type enforceOn = null;
+    List<String> entities = null;
```
No need to assign to `null`. Also all fields of this class should be private final.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,555 @@
+    AuthEnforceAnnotationNodeProcessor(AnnotationNode annotationNode) {
+      List values = annotationNode.values;
+      // AuthEnforce AnnotationNode will have 6 values: 3 field names and another 3 field containing their values.
+      Preconditions.checkArgument(values.size() == 6, String.format("AuthEnforce annotation has three field and " +
```
Preconditions support string formatting already, no need to use `String.format`. Simply do

```java
Preconditions.checkArgument(...., "Need three fields, but found %s", values.size());
```

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,555 @@
+          // class and String[1] is the action which was specified in the annotation
+          List<String[]> actionWithClassName = (List<String[]>) values.get(i + 1);
+          for (String[] action : actionWithClassName) {
+            Preconditions.checkArgument(action.length == 2);
```
Strictly speaking, you should also verify the class name in `action[0]` is the same as the `Action` class name.

rohitsinha54 replied:
Is it necessary ? Since the annotation defines Action[] and passing anything else will be compile time failure. Also, all these methods are private so can't be invoked directly from other class with incorrect data.

chtyim replied:
This is to guard against changes in the `AuthEnforce` annotation.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,555 @@
+        // store for second pass in which we will perform class rewrite
+        @Override
+        public void visitEnd() {
+          if (hasAuthEnforce) {
```
flip the condition

```java
if (authEnforceAnnotationNode == null) {
  return;
}
```

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,555 @@
+          // TODO: Should also pick up names from QueryParam and PathParam annotations here as needed later
+          AnnotationNode annotationNode = new AnnotationNode(desc);
+          // store the parameter position and its annotation detail
+          parameterAnnotationNode.put(parameter, annotationNode);
```
What if the parameter has more than one annotation? E.g.

```java
void fooBar(@PathParam("id") @DefaultValue("xyz") String id)
```

Also, it might contains multiple annotations that are both recognized as "name" by the rewriter (@Name("n1") @PathParam("id") String id). This mean we should have a precedence for those annotations as well.

rohitsinha54 replied:
We will be adding support for QueryParam and PathParam in upcoming PR. I added a note in the existing TODO for the precedence stuff.

chtyim reviewed (commented):
Reviewed the first pass logic. Please address the comments first as it has some edge cases not handled. Will continue the second pass logic.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,550 @@
+    // AuthEnforce.
+    AuthEnforceAnnotationVisitor classVisitor = new AuthEnforceAnnotationVisitor(className);
+    cr.accept(classVisitor,
+              ClassReader.SKIP_CODE | ClassReader.SKIP_DEBUG | ClassReader.SKIP_FRAMES);
```
Seems like you can have it on one line.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,550 @@
+        }
+
+        @Override
+        public AnnotationVisitor visitParameterAnnotation(final int parameter, String desc, boolean visible) {
```
no need to have `final`.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,550 @@
+            // method parameters unless its been specified with this. in that case its always looked in class field.
+            Preconditions.checkArgument(fieldDetails.containsKey(name.startsWith("this.") ?
+                                                                   getFieldName(name) : name) ||
+                                          paramAnnotation.containsKey(name),
```
This condition and indentation is really hard to read. Better separate out to two lines. Also, `Preconditons` support string formatting already, no need to use `String.format`. Moreover, it's better to have `getFieldName` not to throw exception if the name doesn't start with `this.`, but rather have the `? :` logic there. In fact, the condition you want to check should be something like this:

```java
String getFieldName(String paramName) {
  return paramName.startsWith("this.") ? paramName.subString("this.".length()) : name;
}

...
// If a name is prefixed with "this.", have to use field (just like in java if you write "this.x", x has to be a field
boolean valid = fieldDetails.containsKey(getFieldName(name)) 
                     || (!name.startsWith("this.") && paramAnnotation.containsKey(name))
Precondition(valid, "....", ....);
```

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,550 @@
+    Map<Method, AnnotationDetail> getMethodAnnotations() {
+      return methodAnnotations;
+    }
+    Map<String, Type> getFieldDetails() {
```
missing empty line

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,550 @@
+        @Override
+        protected void onMethodEnter() {
+          // Get the annotation details which was collected by visiting annotations during the first pass
+          AnnotationDetail annotationDetail = methodAnnotations.get(new Method(methodName, methodDesc));
```
No need to new the `Method` again. In fact, just do ` get` and check if it is `null` up there.

```java
final AnnotationDetail annotationDetail = methodAnnotations.get(new Method(methodName, methodDesc));
if (annotationDetail == null) {
  return mv;
}
....
```

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,550 @@
+            loadThis();
+            getField(classType, fieldNameWithoutThis, fieldDetails.get(fieldNameWithoutThis));
+          } else if (annotationDetail.getParameterAnnotation().containsKey(name)) { // method parameter has preference
+            // + 1 because 0 specify "this"
```
the comment is outdated.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,550 @@
+          // pushed the entity id
+          //TODO: This is because currently we don't support multi-part
+          String name = annotationDetail.getEntities().get(0);
+          if (name.startsWith("this.")) { // if the entity name was specified with this. then we only look for it in
```
This logic could be simpler if we memorize the decision in the annotation detail, since in the first pass we already performed the validation logic, which means we already knew whether it is a field or a method param. We can even strip the `this.` there.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java:
```diff
@@ -0,0 +1,550 @@
+
+    @Override
+    public void visitEnd() {
+      if (!methodAnnotations.isEmpty()) {
```
This check is not necessary right? Because this 2nd pass won't get triggered if the method annotations is empty, right?

rohitsinha54 commented on cdap-master/src/main/java/co/cask/cdap/data/runtime/main/MasterServiceMain.java:
```diff
@@ -183,8 +185,18 @@
   }
 
   public static void main(final String[] args) throws Exception {
```
This will also initiate the DatasetClassRewriter rewrite is that fine or should we change MainClassLoader to accept what specific rewrite needs to be performed ?

chtyim replied:
That's ok.

chtyim commented:
LGTM

rohitsinha54 commented:
Final Build: http://builds.cask.co/browse/CDAP-DUT5031-23 
ITM: http://builds.cask.co/browse/CDAP-ITM-61
